staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers_test.go

```diff
@@ -0,0 +1,303 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package responsewriters
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/diff"
+	"k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+)
+
+func TestSerializeObject(t *testing.T) {
+	smallPayload := []byte("{test-object,test-object}")
+	largePayload := bytes.Repeat([]byte("0123456789abcdef"), defaultGzipThresholdBytes/16+1)
+	tests := []struct {
+		name string
+
+		compressionEnabled bool
+
+		mediaType  string
+		out        []byte
+		outErrs    []error
+		req        *http.Request
+		statusCode int
+		object     runtime.Object
+
+		wantCode    int
+		wantHeaders http.Header
+		wantBody    []byte
+	}{
+		{
+			name:        "serialize object",
+			out:         smallPayload,
+			req:         &http.Request{Header: http.Header{}},
+			wantCode:    http.StatusOK,
+			wantHeaders: http.Header{"Content-Type": []string{""}},
+			wantBody:    smallPayload,
+		},
+
+		{
+			name:        "return content type",
+			out:         smallPayload,
+			mediaType:   "application/json",
+			req:         &http.Request{Header: http.Header{}},
+			wantCode:    http.StatusOK,
+			wantHeaders: http.Header{"Content-Type": []string{"application/json"}},
+			wantBody:    smallPayload,
+		},
+
+		{
+			name:        "return status code",
+			statusCode:  http.StatusBadRequest,
+			out:         smallPayload,
+			mediaType:   "application/json",
+			req:         &http.Request{Header: http.Header{}},
+			wantCode:    http.StatusBadRequest,
+			wantHeaders: http.Header{"Content-Type": []string{"application/json"}},
+			wantBody:    smallPayload,
```

Comment:
Why for BadRequest we return an object?

Reply:
I guess I'm missing something, because you do this in couple other tests below too.

Reply:
In this case it's emulating a custom registry endpoint (aggregated api) that returns a valid object but with bad request, which is technically possible.  I wanted to have a test that captured behavior that is infrequently used (error code with valid object), but which can show up in some cases (if you had a custom response type that implemented APIStatus, you could return an object + an error code today).

Reply:
Thanks for explanation - that makes a lot of sense to me now.